spring-integration-core/src/main/java/org/springframework/integration/handler/DelayHandler.java

```diff
@@ -365,9 +367,16 @@ private void doReleaseMessage(Message<?> message) {
 
 	private boolean removeDelayedMessageFromMessageStore(Message<?> message) {
 		if (this.messageStore instanceof SimpleMessageStore) {
-			SimpleMessageGroup messageGroup =
-					(SimpleMessageGroup) this.messageStore.getMessageGroup(this.messageGroupId);
-			return messageGroup.remove(message);
+			synchronized (this.messageGroupId) {
+				Collection<Message<?>> messages = this.messageStore.getMessageGroup(this.messageGroupId).getMessages();
+				if (messages.contains(message)) {
+					this.messageStore.removeMessageFromGroup(this.messageGroupId, message);
```

Comment:
Will be changed. But we definitely use it here just only for the `SimpleMessageStore`.
Let's deprecate it in the separate JIRA!
Having such a big change here I even not sure that we will be able to back port anything from here...